src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -1698,59 +1698,16 @@ void Lowering::ContainCheckBinary(GenTreeOp* node)
     CheckImmedAndMakeContained(node, op2);
 
 #ifdef TARGET_ARM64
-    if (comp->opts.OptimizationEnabled() && varTypeIsIntegral(node) && !node->isContained())
+    if (comp->opts.OptimizationEnabled() && !node->isContained() && varTypeIsIntegral(node))
     {
-        // Find "a * b + c" or "c + a * b" in order to emit MADD/MSUB
-        if (node->OperIs(GT_ADD) && !node->gtOverflow() && (op1->OperIs(GT_MUL) || op2->OperIs(GT_MUL)))
+        // Find "a + b * c" or "a - b * c".
+        // Then mark "b * c" op as contained in order to emit 'madd' or 'msub' respectively.
+        if (node->OperIs(GT_ADD, GT_SUB) && !(node->gtFlags & GTF_SET_FLAGS) && !node->gtOverflow() &&
```

Comment:
I think I should make some helper functions to handle this; there are a lot of conditions here...